### PR TITLE
Hog Descriptor Generation

### PR DESCRIPTION
**Pull request by RossBrunton**
This is good enough for us all to review. I guess once we have handed in our report (aiming to do so on Thursday), we should all sit down and read through all of this, make comments, and then presumably fight over minor things.

And I guess since this is a big thing, we should get two "ship its" before it is merged. Also, this includes changes from #11 , so that will need to be merged before this.

... And why can't I add myself as a reviewer? How bothersome.

### CODE REVIEW

File: ImageProcessing/src/image/hog_person_detector/hist_binning_input.cal [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package image.hog_person_detector;
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/hist_binning_input.cal]: Should this file even be here?

File: ImageProcessing/src/image/hog_person_detector/hist_binning_input2.cal [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package image.hog_person_detector;
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/hist_binning_input2.cal]: If these are all for testing, shouldn't they be in `hog_person_detector.test`?

File: ImageProcessing/src/image/hog_person_detector/histogram_binning.cal [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		end
+		i:= 0;
+	end
+
+	bin: action angle:[angle], intensity:[intense] ==>
+	guard
+		read_count < 64
+	do
+		i := intense;
+		if angle <= 10 then
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/histogram_binning.cal]: Shouldn't values less than 10 cause part of it to go into the largest bin?

File: ImageProcessing/src/image/hog_person_detector/histogram_binning.cal [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+			histogram[0] := histogram[0] + i;
+		else
+			if angle >= 170 then
+				histogram[8] := histogram[8] + i;
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/histogram_binning.cal]: And ditto with the smallest bin?

File: ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		histogram_count := histogram_count + 1;
+	end
+
+	transition: action ==>
+	guard
+		histogram_count = histogram_bins
+	do
+		histogram_count := 0;
+	end //normalising all values in the block between 0 and 1 depending on the minimum and maximum values within the whole block
+	write: action ==> Gout:[ out ]
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal]: Probably add an empty line before this action for clarity

File: ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	do //Normalise between 0 and 1
+		if (min - max) = 0 then
+			out := 255;
+		else
+			out := ((in_histogram[histogram_count] - min) * 255) / ((max - min));
+		end
+		histogram_count := histogram_count + 1;
+	end
+
+	reset: action ==>
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal:91]: Doesn't reset max and min.

File: ImageProcessing/src/image/hog_person_detector/inverseslidingwindow.cal [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	end
+
+	write: action ==> Gout:[ out ]
+	guard
+		head < bufferSize
+	do
+
+
+		out := buffer[(y + yBase) * imageWidth + x + xBase];
+		// for x to windowDimension
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/inverseslidingwindow.cal]: Lots of commented out "dead" code

File: ImageProcessing/src/image/hog_person_detector/inverseslidingwindow.cal [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+		// if y >= windowDimension
+			// then xBase ++ windowDinsion
+			// x == 0
+			// y == 0
+			//if xBse >= imageWidth
+				// then yBase++
+				// xbase == 0
+		x := x + 1;
+		if x >= windowDimension then
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/inverseslidingwindow.cal]: Indenting is a bit wacky here

File: ImageProcessing/src/image/hog_person_detector/inversetiler.cal [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+/*
+ * Author: Mark Young & Teymoor Rasheed Ali my120@hw.ac.uk tr78@hw.ac.uk
+ * Date: 27.02.2017
+ */
+package image.hog_person_detector;
+
+actor inversetiler () uint(size=8) Gin ==> uint(size=8) Gout :
+
+/* Constants */
+	uint(size=8) imageHeight = 128;
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/inversetiler.cal:41]: Forgot to update this to use the constants, whoops!

File: ImageProcessing/src/image/hog_person_detector/tiler.cal [hunk 10]
```diff
@@ -0,0 +1,118 @@
+
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/tiler.cal:1]: I think I've seen this before

**Review by RossBrunton — COMMENTED** [review]
Had a look through everything, and left some comments. Some of them aren't about whitespace this time! :D

(Obligatory I suck at tone, so don't take my abusive comments to heart or whatever)

File: ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal (hunk 6 above)
Comment (Teymoor-Ali) [reply ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal:91]: Yep saw that issue, ill push an update through

File: ImageProcessing/src/image/hog_person_detector/inverseslidingwindow.cal (hunk 7 above)
Comment (Teymoor-Ali) [reply ImageProcessing/src/image/hog_person_detector/inverseslidingwindow.cal]: dont need inversesliding window or sliding window,  removing those files

Comment (Teymoor-Ali) [reply ImageProcessing/src/image/hog_person_detector/inverseslidingwindow.cal]: tiler/inverse tiler are the same thing

File: ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal [hunk 11]
```diff
@@ -35,12 +35,10 @@
 */
 package image.hog_person_detector;
 
-import image.hog_person_detector.constants.*;
-
 actor histogram_normalisation () uint(size=16) Gin ==> uint(size=8) Gout :
 
 /* Constants */
-	uint(size=8) histogram_bins := HIST_LENGTH * 4;
+	uint(size=8) histogram_bins := 36;
```
Comment (RossBrunton) [inline ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal]: Why has this been reverted?

Comment (Teymoor-Ali) [reply ImageProcessing/src/image/hog_person_detector/histogram_normalisation.cal]: oh must of copy and pasted the old one, ill fix it.

**Comment by RossBrunton** [comment]
Calum isn't able to work on it until tomorrow, so asked me to do his changes. Unfortunately I changed a lot of things which appeared to be bugs to me, but may not be. Calum, when you get back could you give the changes a look over and run the tests to make sure everything works?

**Review by Teymoor-Ali — COMMENTED** [review]
Looks good,

**Review by cajturner — APPROVED** [review]
Changes look fine. I wasn't aware that the bins wrapped around. Test return expected output

**Comment by robstewart57** [comment]
@cajturner @RossBrunton so... can I squash and merge this PR now?

**Comment by RossBrunton** [comment]
I think so, all my review comments seem to be addressed.

**Pull request merged**